### PR TITLE
raftstore: add RequestInspector to decide how to handle requests

### PR DESCRIPTION
**Pull request by overvenus**
What have you changed? (mandatory)

Add a struct to decouple `get_handle_policy` from `Peer`, #2639  needs it to handle requests.

Refer to a related PR or issue link (optional)

#2639

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1875,6 +1800,123 @@ impl Peer {
...
+
+impl<'a, T: RequestJury> RequestJudge<'a, T> {
+    /// Create a `RequestJudge`.
+    fn with_jury(jury: &'a T) -> RequestJudge<T> {
+        RequestJudge { jury }
+    }
+
+    /// Judges a request, return a doctrine that tells us how to
+    /// handle the request.
+    fn judge(&self, req: &RaftCmdRequest) -> Result<RequestDoctrine> {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: Please add some tests.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -273,7 +263,7 @@ pub struct Peer {
 
     leader_missing_time: Option<Instant>,
 
-    leader_lease: Lease,
+    leader_lease: RefCell<Lease>,
```
Comment (huachaohuang) [inline src/raftstore/store/peer.rs]: Why add `RefCell` here?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: I need its interior mutability at line https://github.com/pingcap/tikv/pull/3193/files#diff-4fe7d196e1e0c8626e9abc9ffdb98ef2R1912

**Comment by overvenus** [comment]
> Are those names RequestJudge, RequestJury, RequestDoctrine in this PR too funny?

Why funny? They are serious. :classical_building: 

> Should we use more common words like RequestInspector, RequestExecutor, and RequestType?

`RequestInspector`, I think it does not common neither. 
`RequestExecutor`, I don't think it suits, because it does not execute requests at this moment.
`RequestType`, it's not suit too, because they are NOT the type of requests but the way how we should handle requests.

I am willing to change if there are better name (I has a poor vocabulary :joy:)

**Comment by BusyJay** [comment]
I think `Inspector` is a very common pattern. `Judge`, `Jury` and `Doctrine` feels too meaningful to me. I need extra knowledge to fully understand what are they and how do they work (maybe a 10 minutes read of Wikipedia pages). Use the more verbose and common used words is better.

**Comment by overvenus** [comment]
Okay, you win.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1875,6 +1800,123 @@ impl Peer {
...
+}
+
+/// Judge makes `RequestDoctrine` for requests.
+struct RequestJudge<'a, T: 'a + RequestJury> {
+    jury: &'a T,
+}
+
+impl<'a, T: RequestJury> RequestJudge<'a, T> {
+    /// Create a `RequestJudge`.
+    fn with_jury(jury: &'a T) -> RequestJudge<T> {
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: You can use `impl RequestJury` here now I think. :)

Comment (overvenus) [reply src/raftstore/store/peer.rs]: Well, I think `&'a T` is fine, I like the idea of [the RFC]( https://github.com/phaazon/rfcs/blob/undo-universal-impl-trait/text/0000-undo-universal-impl-trait.md#why-is-impl-trait-in-argument-position-wrong) .

Comment (Hoverbear) [reply src/raftstore/store/peer.rs]: Ok :)

**Comment by Hoverbear** [comment]
Aww I liked the `Jury`/`Judge` stuff. :)

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -1875,6 +1800,124 @@ impl Peer {
...
+    ReadLocal,
+    // Handle the read request via raft's SafeReadIndex mechanism.
+    ReadIndex,
+    ProposeNormal,
+    ProposeTransferLeader,
+    ProposeConfChange,
+}
+
+/// A `RequestCondition` helps `RequestInspector` make a `RequestPolicy`.
+trait RequestCondition {
```
Comment (huachaohuang) [inline src/raftstore/store/peer.rs]: Sorry, I can't connect the name "condition" with these methods, how about `RequestHandler`?

Comment (huachaohuang) [reply src/raftstore/store/peer.rs]: What will implement this trait?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: `Peer`, but it does not mean this trait is supposed to to execute or handle requests. You know traits are independent to their implementer. I can define a new struct who implements the trait.

```rust
struct Foo {
  appiled_index_term,
  lease,
}
```

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -1882,6 +1807,111 @@ impl Peer {
     }
 }
 
+/// `RequestPolicy` decides how we handle a request.
+#[derive(Clone, PartialEq, Debug)]
+enum RequestPolicy {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Should it be `pub`?

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -1882,6 +1807,111 @@ impl Peer {
...
+                    ));
+                }
+            }
+
+            if has_read && has_write {
+                return Err(box_err!("read and write can't be mixed in one batch."));
+            }
+        }
+
+        if has_write {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Seems `is_write` is more readable.

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -1882,6 +1807,111 @@ impl Peer {
...
+            LeaseState::Expired | LeaseState::Suspect => {
+                // Perform a consistent read to Raft quorum and try to renew the leader lease.
+                Ok(RequestPolicy::ReadIndex)
+            }
+        }
+    }
+}
+
+impl RequestInspector for Peer {
+    fn has_applied_current_term(&self) -> bool {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: It may not be accurate. As it has not applied all logs from current term.

File: src/raftstore/store/peer.rs [hunk 8]
```diff
@@ -1882,6 +1807,111 @@ impl Peer {
...
+        self.get_store().applied_index_term == self.term()
+    }
+
+    fn inspect_lease(&mut self) -> LeaseState {
+        if !self.raft_group.raft.in_lease() {
+            return LeaseState::Suspect;
+        }
+        // None means now.
+        let state = self.leader_lease.inspect(None);
+        match state {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: How about `if let`?

File: src/raftstore/store/peer.rs [hunk 9]
```diff
@@ -2055,4 +2085,139 @@ mod tests {
...
+            (CmdType::DeleteRange, RequestPolicy::ProposeNormal),
+            (CmdType::IngestSST, RequestPolicy::ProposeNormal),
+        ] {
+            let mut request = raft_cmdpb::Request::new();
+            request.set_cmd_type(op);
+            req.set_requests(vec![request].into());
+            table.push((req.clone(), policy));
+        }
+
+        for b in vec![true, false] {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: How about list them explicitly? Seems only list the read requests is enough.

File: src/raftstore/store/peer.rs (hunk 8 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: How about other states? I want it returns state precisely.

File: src/raftstore/store/peer.rs (hunk 7 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: What's your suggestion?

File: src/raftstore/store/peer.rs (hunk 9 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: I think it's fine. It also improves coverage.

File: src/raftstore/store/peer.rs (hunk 8 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: `if let` just do special things for certain state. And then always return the `state` itself.

File: src/raftstore/store/peer.rs (hunk 9 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: How can they improve coverage as they always hit the same line? List them explicitly can make tests more clear, so reader don't need to think if the transformation rule is correct or not.

File: src/raftstore/store/peer.rs (hunk 7 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: How about `has_applied_to_current_term`?

File: src/raftstore/store/peer.rs (hunk 9 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: After listing them one by one, the tests becoms to long, instead, I will add some comments to make it more clear.

**Pull request merged**